doc/man/gnunet-ecc.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd June 5, 2017
.Dt GNUNET-ECC 1
.Os
.Sh NAME
.Nm gnunet-ecc
.Nd manipulate GNUnet ECC key files
.Sh SYNOPSIS
.Nm
.Op Fl g Ar COUNT | Fl -generate-keys= Ns Ar COUNT
.Op Fl p | -print-public-key
.Op Fl P | -print-private-key
.Op Fl x | -print-hex
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl v | -version
.Ao Ar FILENAME Ac
.sp
.Sh DESCRIPTION
.Nm
can be used to create an ECC private key and to print the corresponding public key.
You must specify a filename containing an ECC private key in GNUnet format as an argument.
If the file does not exist, gnunet-ecc will create a key.
This may then take a while.
If the option
.Fl p
is given, the corresponding public key will be printed to the console.
.Bl -tag -width indent
.It Fl g Ar COUNT | Fl -generate-keys= Ns Ar COUNT
Create COUNT public-private key pairs and write them to FILENAME.
Used for creating a file for testing.
.It Fl p | -print-public-key
Print the corresponding public key to stdout.
This is the value used for PKEY records in GNS.
.It Fl P | -print-private-key
Print the corresponding private key to stdout.
This is the value used for PKEY records in GNS.
.It Fl x | -print-hex
Print the corresponding public key to stdout in HEX format.
Useful for comparing to Ed25519 keys in X.509 tools.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl v | -version
Print GNUnet version number.
.El
.Sh SEE ALSO
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression
  in 0.11.0 where documentation was an on/off switch.
  New introduced: --enable-include-manpages.
  Fixed: texi2mdoc integration (--enable-texi2mdoc-generation).
  buildsystem: decrease python dependency to 3.4 from
  previously 3.7.
  -ng0

Tue Mar 12 10:06:17 CET 2019
  REST: Move back from rest-plugin into respective subsystem directories.

Mon Mar 11 02:47:55 CET 2019
  Make configure complain if no working libidn is detected. -CG

Sat Mar 9 15:58:45 2019 +0100
  REST: Config option for address bind. Defaults to localhost.
  GNS: dns2gns/gns-proxy config option for address bind. Defaults to localhost.

Sat Mar 9 01:58:22 CET 2019
  gnunet-publish now by default does not expose the creation time,
  option -d no longer exists, to enable the old behavior use -E. -CG

Sat Mar 2 14:23:10 2019 +0100

src/util/gnunet-ecc.c
/*
     This file is part of GNUnet.
     Copyright (C) 2012, 2013 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file util/gnunet-ecc.c
 * @brief tool to manipulate EDDSA key files
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_testing_lib.h"
#include <gcrypt.h>

/**
 * Number of characters a Base32-encoded public key requires.
 */
#define KEY_STR_LEN sizeof(struct GNUNET_CRYPTO_EddsaPublicKey) * 8 / 5 + 1

/**
 * Flag for listing public key.
 */
static int list_keys;

/**
 * Flag for listing public key.
 */
static unsigned int list_keys_count;

/**
 * Flag for printing public key.
 */
static int print_public_key;

/**
 * Flag for printing private key.
 */
static int print_private_key;

/**
 * Flag for printing public key in hex.
 */
static int print_public_key_hex;

/**
 * Flag for printing the output of random example operations.
 */
static int print_examples_flag;

/**
 * Option set to create a bunch of keys at once.
 */
static unsigned int make_keys;


/**
 * Create a flat file with a large number of key pairs for testing.
 *
 * @param fn File name to store the keys.
 * @param prefix Desired prefix for the public keys, NULL if any key is OK.
 */
static void
create_keys (const char *fn, const char *prefix)
{
  FILE *f;
  struct GNUNET_CRYPTO_EddsaPrivateKey *pk;
  struct GNUNET_CRYPTO_EddsaPublicKey target_pub;
  static char vanity[KEY_STR_LEN + 1];
  size_t len;
  size_t n;
  size_t rest;
  unsigned char mask;
  unsigned target_byte;
  char *s;

  if (NULL == (f = fopen (fn, "w+")))
  {
    fprintf (stderr, _ ("Failed to open `%s': %s\n"), fn, strerror (errno));
    return;
  }
  if (NULL != prefix)
  {
    len = GNUNET_strlcpy (vanity, prefix, sizeof(vanity));
    n = len * 5 / 8;
    rest = len * 5 % 8;

    memset (&vanity[len], '0', KEY_STR_LEN - len);
    vanity[KEY_STR_LEN] = '\0';
    GNUNET_assert (GNUNET_OK ==
                   GNUNET_CRYPTO_eddsa_public_key_from_string (vanity,
                                                               KEY_STR_LEN,
                                                               &target_pub));
    if (0 != rest)
    {
      /**
       * Documentation by example:
       * vanity = "A"
       * len = 1
       * n = 5/8 = 0 (bytes)
       * rest = 5%8 = 5 (bits)
       * mask = ~(2**(8 - 5) - 1) = ~(2**3 - 1) = ~(8 - 1) = ~b111 = b11111000
       */
      mask = ~((int) pow (2, 8 - rest) - 1);
      target_byte = ((unsigned char *) &target_pub)[n] & mask;
    }
    else
    {
      /* Just so old (debian) versions of GCC calm down with the warnings. */
      mask = target_byte = 0;
    }
    s = GNUNET_CRYPTO_eddsa_public_key_to_string (&target_pub);
    fprintf (stderr,
             _ ("Generating %u keys like %s, please wait"),
             make_keys,
             s);
    GNUNET_free (s);
    fprintf (stderr, "\nattempt %s [%u, %X]\n", vanity, (unsigned int) n, mask);
  }
  else
  {
    fprintf (stderr, _ ("Generating %u keys, please wait"), make_keys);
    /* Just so old (debian) versions of GCC calm down with the warnings. */
    n = rest = target_byte = mask = 0;
  }

  while (0 < make_keys--)
  {
    fprintf (stderr, ".");
    if (NULL == (pk = GNUNET_CRYPTO_eddsa_key_create ()))
    {
      GNUNET_break (0);
      break;
    }
    if (NULL != prefix)
    {
      struct GNUNET_CRYPTO_EddsaPublicKey newkey;

      GNUNET_CRYPTO_eddsa_key_get_public (pk, &newkey);
      if (0 != memcmp (&target_pub, &newkey, n))
      {
        make_keys++;
        continue;
      }
      if (0 != rest)
      {
        unsigned char new_byte;

        new_byte = ((unsigned char *) &newkey)[n] & mask;
        if (target_byte != new_byte)
        {
          make_keys++;
          continue;
        }
      }
    }
    if (GNUNET_TESTING_HOSTKEYFILESIZE !=
        fwrite (pk, 1, GNUNET_TESTING_HOSTKEYFILESIZE, f))
    {
      fprintf (stderr,
               _ ("\nFailed to write to `%s': %s\n"),
               fn,
               strerror (errno));
      GNUNET_free (pk);
      break;
    }
    GNUNET_free (pk);
  }
  if (UINT_MAX == make_keys)
    fprintf (stderr, _ ("\nFinished!\n"));
  else
    fprintf (stderr, _ ("\nError, %u keys not generated\n"), make_keys);
  fclose (f);
}


static void
print_hex (const char *msg, const void *buf, size_t size)
{
  printf ("%s: ", msg);
  for (size_t i = 0; i < size; i++)
  {
    printf ("%02hhx", ((const uint8_t *) buf)[i]);
  }
  printf ("\n");
}


static void
print_examples_ecdh ()
{
  struct GNUNET_CRYPTO_EcdhePrivateKey *dh_priv1;
  struct GNUNET_CRYPTO_EcdhePublicKey *dh_pub1;
  struct GNUNET_CRYPTO_EcdhePrivateKey *dh_priv2;
  struct GNUNET_CRYPTO_EcdhePublicKey *dh_pub2;
  struct GNUNET_HashCode hash;
  char buf[128];

  dh_pub1 = GNUNET_new (struct GNUNET_CRYPTO_EcdhePublicKey);
  dh_priv1 = GNUNET_CRYPTO_ecdhe_key_create ();
  dh_pub2 = GNUNET_new (struct GNUNET_CRYPTO_EcdhePublicKey);
  dh_priv2 = GNUNET_CRYPTO_ecdhe_key_create ();
  GNUNET_CRYPTO_ecdhe_key_get_public (dh_priv1, dh_pub1);
  GNUNET_CRYPTO_ecdhe_key_get_public (dh_priv2, dh_pub2);

  GNUNET_assert (NULL !=
                 GNUNET_STRINGS_data_to_string (dh_priv1, 32, buf, 128));
  printf ("ECDHE key 1:\n");
  printf ("private: %s\n", buf);
  print_hex ("private(hex)", dh_priv1, sizeof *dh_priv1);
  GNUNET_assert (NULL != GNUNET_STRINGS_data_to_string (dh_pub1, 32, buf, 128));
  printf ("public: %s\n", buf);
  print_hex ("public(hex)", dh_pub1, sizeof *dh_pub1);

  GNUNET_assert (NULL !=
                 GNUNET_STRINGS_data_to_string (dh_priv2, 32, buf, 128));
  printf ("ECDHE key 2:\n");
  printf ("private: %s\n", buf);
  print_hex ("private(hex)", dh_priv2, sizeof *dh_priv2);
  GNUNET_assert (NULL != GNUNET_STRINGS_data_to_string (dh_pub2, 32, buf, 128));
  printf ("public: %s\n", buf);
  print_hex ("public(hex)", dh_pub2, sizeof *dh_pub2);

  GNUNET_assert (GNUNET_OK ==
                 GNUNET_CRYPTO_ecc_ecdh (dh_priv1, dh_pub2, &hash));
  GNUNET_assert (NULL != GNUNET_STRINGS_data_to_string (&hash, 64, buf, 128));
  printf ("ECDH shared secret: %s\n", buf);

  GNUNET_free (dh_priv1);
  GNUNET_free (dh_priv2);
  GNUNET_free (dh_pub1);
  GNUNET_free (dh_pub2);
}


/**
 * Print some random example operations to stdout.
 */
static void
print_examples ()
{
  print_examples_ecdh ();
  // print_examples_ecdsa ();
  // print_examples_eddsa ();
}


static void
print_key (const char *filename)
{
  struct GNUNET_DISK_FileHandle *fd;
  struct GNUNET_CRYPTO_EddsaPrivateKey private_key;
  struct GNUNET_CRYPTO_EddsaPublicKey public_key;
  char *hostkeys_data;
  char *hostkey_str;
  uint64_t fs;
  unsigned int total_hostkeys;
  unsigned int c;
  ssize_t sret;

  if (GNUNET_YES != GNUNET_DISK_file_test (filename))
  {
    fprintf (stderr, _ ("Hostkeys file `%s' not found\n"), filename);
    return;
  }

  /* Check hostkey file size, read entire thing into memory */
  if (GNUNET_OK !=
      GNUNET_DISK_file_size (filename, &fs, GNUNET_YES, GNUNET_YES))
    fs = 0;
  if (0 == fs)
  {
    fprintf (stderr, _ ("Hostkeys file `%s' is empty\n"), filename);
    return;   /* File is empty */
  }
  if (0 != (fs % GNUNET_TESTING_HOSTKEYFILESIZE))
  {
    fprintf (stderr, _ ("Incorrect hostkey file format: %s\n"), filename);
    return;
  }
  fd = GNUNET_DISK_file_open (filename,
                              GNUNET_DISK_OPEN_READ,
                              GNUNET_DISK_PERM_NONE);
  if (NULL == fd)
  {
    GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "open", filename);
    return;
  }
  hostkeys_data = GNUNET_malloc (fs);
  sret = GNUNET_DISK_file_read (fd, hostkeys_data, fs);
  if ((sret < 0) || (fs != (size_t) sret))
  {
    fprintf (stderr, _ ("Could not read hostkey file: %s\n"), filename);
    GNUNET_free (hostkeys_data);
    GNUNET_DISK_file_close (fd);
    return;
  }
  GNUNET_DISK_file_close (fd);

  if (NULL == hostkeys_data)
    return;
  total_hostkeys = fs / GNUNET_TESTING_HOSTKEYFILESIZE;
  for (c = 0; (c < total_hostkeys) && (c < list_keys_count); c++)
  {
    GNUNET_memcpy (&private_key,
                   hostkeys_data + (c * GNUNET_TESTING_HOSTKEYFILESIZE),
                   GNUNET_TESTING_HOSTKEYFILESIZE);
    GNUNET_CRYPTO_eddsa_key_get_public (&private_key, &public_key);
    hostkey_str = GNUNET_CRYPTO_eddsa_public_key_to_string (&public_key);
    if (NULL != hostkey_str)
    {
      fprintf (stderr, "%4u: %s\n", c, hostkey_str);
      GNUNET_free (hostkey_str);
    }
    else
      fprintf (stderr, "%4u: %s\n", c, "invalid");
  }
  GNUNET_free (hostkeys_data);
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure, NULL
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  (void) cls;
  (void) cfgfile;
  (void) cfg;

  if (print_examples_flag)
  {
    print_examples ();
    return;
  }
  if (NULL == args[0])
  {
    fprintf (stderr, "%s", _ ("No hostkey file specified on command line\n"));
    return;
  }
  if (list_keys)
  {
    print_key (args[0]);
    return;
  }
  if (make_keys > 0)
  {
    create_keys (args[0], args[1]);
    return;
  }
  if (print_public_key || print_public_key_hex || print_private_key)
  {
    char *str;
    struct GNUNET_DISK_FileHandle *keyfile;
    struct GNUNET_CRYPTO_EddsaPrivateKey pk;
    struct GNUNET_CRYPTO_EddsaPublicKey pub;

    keyfile = GNUNET_DISK_file_open (args[0],
                                     GNUNET_DISK_OPEN_READ,
                                     GNUNET_DISK_PERM_NONE);
    if (NULL == keyfile)
      return;
    while (sizeof(pk) == GNUNET_DISK_file_read (keyfile, &pk, sizeof(pk)))
    {
      GNUNET_CRYPTO_eddsa_key_get_public (&pk, &pub);
      if (print_public_key_hex)
      {
        print_hex ("HEX:", &pub, sizeof(pub));
      }
      else if (print_public_key)
      {
        str = GNUNET_CRYPTO_eddsa_public_key_to_string (&pub);
        fprintf (stdout, "%s\n", str);
        GNUNET_free (str);
      }
      else if (print_private_key)
      {
        str = GNUNET_CRYPTO_eddsa_private_key_to_string (&pk);
        fprintf (stdout, "%s\n", str);
        GNUNET_free (str);
      }
    }
    GNUNET_DISK_file_close (keyfile);
  }
}


/**
 * Program to manipulate ECC key files.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag ('i',
                               "iterate",
                               gettext_noop (
                                 "list keys included in a file (for testing)"),
                               &list_keys),
    GNUNET_GETOPT_option_uint (
      'e',
      "end=",
      "COUNT",
      gettext_noop ("number of keys to list included in a file (for testing)"),
      &list_keys_count),
    GNUNET_GETOPT_option_uint (
      'g',
      "generate-keys",
      "COUNT",
      gettext_noop ("create COUNT public-private key pairs (for testing)"),
      &make_keys),
    GNUNET_GETOPT_option_flag ('p',
                               "print-public-key",
                               gettext_noop (
                                 "print the public key in ASCII format"),
                               &print_public_key),
    GNUNET_GETOPT_option_flag ('P',
                               "print-private-key",
                               gettext_noop (
                                 "print the private key in ASCII format"),
                               &print_private_key),
    GNUNET_GETOPT_option_flag ('x',
                               "print-hex",
                               gettext_noop (
                                 "print the public key in HEX format"),
                               &print_public_key_hex),
    GNUNET_GETOPT_option_flag (
      'E',
      "examples",
      gettext_noop (
        "print examples of ECC operations (used for compatibility testing)"),
      &print_examples_flag),
    GNUNET_GETOPT_OPTION_END };
  int ret;

  list_keys_count = UINT32_MAX;
  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  ret = (GNUNET_OK ==
         GNUNET_PROGRAM_run (argc,
                             argv,
                             "gnunet-ecc [OPTIONS] keyfile [VANITY_PREFIX]",
                             gettext_noop (
                               "Manipulate GNUnet private ECC key files"),
                             options,
                             &run,
                             NULL))
        ? 0
        : 1;
  GNUNET_free ((void *) argv);
  return ret;
}

/* end of gnunet-ecc.c */
